File: apps/desktop/src/lib/mesh.test.ts
import { describe, it, expect } from 'vitest'
import { canTransitionMesh, dominatesVector, isVectorConcurrent, mergeVector, isVectorMonotonic, lwwMerge, validateMeshSyncPayload, validateMeshConflictPayload } from './mesh'

describe('mesh state', () => {
  it('transitions', () => {
    expect(canTransitionMesh('IDLE','SYNCING')).toBe(true)
    expect(canTransitionMesh('SYNCING','CONFLICT')).toBe(true)
    expect(canTransitionMesh('CONFLICT','SYNCING')).toBe(true)
    expect(canTransitionMesh('SYNCING','IDLE')).toBe(true)
    expect(canTransitionMesh('CONFLICT','IDLE')).toBe(true)
    expect(canTransitionMesh('IDLE','CONFLICT')).toBe(false)
    expect(canTransitionMesh('IDLE','IDLE')).toBe(false)
  })
  it('vector dominates', () => {
    expect(dominatesVector({a:3,b:2},{a:2,b:2})).toBe(true)
    expect(dominatesVector({a:2,b:2},{a:3,b:2})).toBe(false)
    expect(dominatesVector({a:1},{a:1})).toBe(false)
  })
  it('concurrent', () => {
    expect(isVectorConcurrent({a:3,b:1},{a:2,b:2})).toBe(true)
    expect(isVectorConcurrent({a:3,b:2},{a:2,b:2})).toBe(false)
    expect(isVectorConcurrent({a:3},{a:3})).toBe(false)
  })
  it('merge', () => {
    expect(mergeVector({a:1},{b:2})).toEqual({a:1,b:2})
    expect(mergeVector({a:3,b:1},{a:2,b:5})).toEqual({a:3,b:5})
  })
  it('monotonic', () => {
    expect(isVectorMonotonic({a:2},{a:3},'a')).toBe(true)
    expect(isVectorMonotonic({a:2},{a:5},'a')).toBe(false)
    expect(isVectorMonotonic({a:2},{a:2},'a')).toBe(true)
  })
  it('lww', () => {
    const a = {text:'hello', mime:'text/plain', ts:1000, device_id:'a'}
    const b = {text:'world', mime:'text/plain', ts:2000, device_id:'b'}
    expect(lwwMerge(a,b).text).toBe('world')
    expect(lwwMerge(b,a).text).toBe('world')
    const c = {text:'aaa', mime:'text/plain', ts:2000, device_id:'a'}
    const d = {text:'bbb', mime:'text/plain', ts:2000, device_id:'b'}
    expect(lwwMerge(c,d).device_id).toBe('b')
  })
  it('validate mesh sync ok', () => {
    expect(validateMeshSyncPayload({deviceId:'phone-xyz', vectors:{'phone-xyz':1}, entries:[{path:'/report.pdf', vector:{'phone-xyz':1}}], ts: Date.now()})).toBe(null)
  })
  it('validate mesh sync missing device', () => {
    expect(validateMeshSyncPayload({vectors:{}, entries:[]})).not.toBe(null)
  })
  it('validate mesh sync invalid path', () => {
    expect(validateMeshSyncPayload({deviceId:'phone-xyz', vectors:{}, entries:[{path:'../bad'}]})).not.toBe(null)
  })
  it('validate mesh conflict ok', () => {
    expect(validateMeshConflictPayload({path:'/report.pdf', resolution:'lww', winner:'local'})).toBe(null)
    expect(validateMeshConflictPayload({path:'/report.pdf', resolution:'rename', winner:'remote', loserRename:'/report.pdf.conflict'})).toBe(null)
  })
  it('validate mesh conflict invalid', () => {
    expect(validateMeshConflictPayload({path:'/a', resolution:'bad', winner:'local'})).not.toBe(null)
    expect(validateMeshConflictPayload({path:'../a', resolution:'lww', winner:'local'})).not.toBe(null)
  })
})

File: crates/bridge-core/tests/mesh_test.rs
use bridge_core::{MeshState, vector_clock_dominates, is_vector_concurrent, vector_clock_merge, LwwClipboard, lww_clipboard_merge, validate_mesh_sync_payload, validate_mesh_conflict_payload};
use serde_json::json;

#[test]
fn mesh_state_valid() {
    assert!(MeshState::Idle.can_transition(&MeshState::Syncing));
    assert!(MeshState::Syncing.can_transition(&MeshState::Conflict));
    assert!(MeshState::Conflict.can_transition(&MeshState::Syncing));
    assert!(MeshState::Syncing.can_transition(&MeshState::Idle));
    assert!(MeshState::Conflict.can_transition(&MeshState::Idle));
}
#[test]
fn mesh_state_invalid() {
    assert!(!MeshState::Idle.can_transition(&MeshState::Conflict));
    assert!(!MeshState::Idle.can_transition(&MeshState::Idle));
    assert!(!MeshState::Conflict.can_transition(&MeshState::Conflict));
}
#[test]
fn mesh_vector_dominates() {
    let mut a = std::collections::HashMap::new();
    a.insert("desktop".into(), 3);
    a.insert("phone".into(), 2);
    let mut b = std::collections::HashMap::new();
    b.insert("desktop".into(), 2);
    b.insert("phone".into(), 2);
    assert!(vector_clock_dominates(&a,&b));
    assert!(!vector_clock_dominates(&b,&a));
}
#[test]
fn mesh_vector_concurrent() {
    let mut a = std::collections::HashMap::new();
    a.insert("desktop".into(), 3);
    a.insert("phone".into(), 1);
    let mut b = std::collections::HashMap::new();
    b.insert("desktop".into(), 2);
    b.insert("phone".into(), 2);
    assert!(is_vector_concurrent(&a,&b));
}
#[test]
fn mesh_merge() {
    let mut a = std::collections::HashMap::new(); a.insert("desktop".into(), 1);
    let mut b = std::collections::HashMap::new(); b.insert("phone".into(), 2);
    let m = vector_clock_merge(&a,&b);
    assert_eq!(m.get("desktop"), Some(&1));
    assert_eq!(m.get("phone"), Some(&2));
}
#[test]
fn mesh_lww_merge() {
    let a = LwwClipboard{text:"hello".into(), mime:"text/plain".into(), ts:1000, device_id:"a".into()};
    let b = LwwClipboard{text:"world".into(), mime:"text/plain".into(), ts:2000, device_id:"b".into()};
    assert_eq!(lww_clipboard_merge(&a,&b).text, "world");
    assert_eq!(lww_clipboard_merge(&b,&a).text, "world");
    let c = LwwClipboard{text:"aaa".into(), mime:"text/plain".into(), ts:2000, device_id:"a".into()};
    let d = LwwClipboard{text:"bbb".into(), mime:"text/plain".into(), ts:2000, device_id:"b".into()};
    assert_eq!(lww_clipboard_merge(&c,&d).device_id, "b");
}
#[test]
fn validate_mesh_sync_ok() {
    let p = json!({"deviceId":"phone-xyz","vectors":{"phone-xyz":1},"entries":[{"path":"/report.pdf","vector":{"phone-xyz":1}}],"ts":chrono::Utc::now().timestamp_millis()});
    assert!(validate_mesh_sync_payload(&p).is_ok());
}
#[test]
fn validate_mesh_sync_missing_device() {
    let p = json!({"vectors":{},"entries":[]});
    assert!(validate_mesh_sync_payload(&p).is_err());
}
#[test]
fn validate_mesh_sync_bad_path() {
    let p = json!({"deviceId":"phone-xyz","vectors":{},"entries":[{"path":"../bad"}]});
    assert!(validate_mesh_sync_payload(&p).is_err());
}
#[test]
fn validate_mesh_conflict_ok() {
    let p = json!({"path":"/report.pdf","resolution":"lww","winner":"local","loserRename":"/report.pdf.mesh-conflict"});
    assert!(validate_mesh_conflict_payload(&p).is_ok());
}
#[test]
fn validate_mesh_conflict_invalid() {
    let p = json!({"path":"/a","resolution":"bad","winner":"local"});
    assert!(validate_mesh_conflict_payload(&p).is_err());
    let p2 = json!({"path":"../a","resolution":"lww","winner":"local"});
    assert!(validate_mesh_conflict_payload(&p2).is_err());
}
